### PR TITLE
Ruby Enumerable Methods

### PR DESCRIPTION
**Pull request by frankopkusianwar**
#### What does this PR do?

- Adds Ruby enumerable methods to the enumerable module

### Description of Task to be completed?

 - Added the enumerable_custom.rb file.
 - Created the enumerable class and added the enumerable methods 
 - Ran validations on Rubocop.

#### How should this be manually tested?

- Clone this branch.
- open the enumerables_custom.rb on your code editor.
- open the terminal and run  *ruby enumerable_custom.rb* command

#### Screenshot

screenshot

### CODE REVIEW

File: enumerable_custom.rb [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+  def my_select(*)
+    if block_given?
+      array = []
+      my_each { |item| array << item if yield(item) }
+    else
+      to_enum(:my_select)
+    end
+    array
+  end
```
Comment (stratospherique) [inline enumerable_custom.rb]: * [x] should have working `my_select` identical and returns the same thing as ruby's [`select`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-select). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_select.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

* [x] my_select returns an enumerator if no block is given
  ```
    array = [1,2,5,8,7,98,4]
    p array.my_select.class # not Enumerator
  ```

File: enumerable_custom.rb [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+      to_enum(:my_any?)
+    end
+    condition
+  end
+
+  def my_none?
+    condition = true
+    my_each { |item| condition = false if yield(item) }
+    condition
+  end
```
Comment (stratospherique) [inline enumerable_custom.rb]: The issues with **my_none?**, **my_any?** and **my_all?** are similar. the methods should be able to: 

* [x] Return a suitable boolean result if no arguments were passed.
  ```ruby
	array = [false, true, false, nil, []]
	puts array.my_any? # should print true
	puts array.my_all? # should print false
	puts array.my_none? # should print false
  ```
* [x] Accept primitives, classes (i.e Integer, String, etc...) and Regex Expressions as arguments.
  ```ruby
	words = %w[dog door rod blade]
	puts words.my_all?(/c/) == words.all?(/c/) # false
	puts words.my_any?(/c/) == words.all?(/c/) # false
	puts words.my_none?(/c/) == words.all?(/c/) # true
  ```

File: enumerable_custom.rb [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+      map = []
+      if proc
+        my_each { |item| map << proc.call(item) }
+      else
+        map << yield(item)
+      end
+    else
+      to_enum(:my_map)
+    end
+  end
```
Comment (stratospherique) [inline enumerable_custom.rb]: * [x] should have working `my_map` identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

* [x] my_map returns a new array with the results of running block once for every element in enum.
  ```ruby
    array = [1,2,5,4,7]
    block = proc { |num| num < 5 }
    p array.my_map?(&block) == array.map?(&block) # false
  ```

File: enumerable_custom.rb [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+      sum = start
+      my_each { |item| sum = yield(sum, item) }
+      sum
+    else
+      if start.is_a?(Symbol)
+        my_each { |item| reducer = reducer ? reducer.send(start, item) : item }
+      end
+      reducer
+    end
+  end
```
Comment (stratospherique) [inline enumerable_custom.rb]: * [x] should have working `my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

* [x] my_inject combines all elements of enum by applying a binary operation, specified by a block
  ```ruby
    array = [4,7,8,89,9,6,5,12,4,2,5]
    operation = proc { |sum, n| sum + n  }
    p array.my_inject(&operation) == array.inject(&operation) # false
  ```

* [x] my_inject when a symbol is specified combines each element of the collection by applying the symbol as a named method
  ```ruby
    range = Range.new(5, 50)
    p range.my_inject(2, :*) # returns this error
    # ArgumentError:
        wrong number of arguments (given 2, expected 0..1)
      # ./enumerable_custom.rb:99:in `my_inject'
  ```

**Review by stratospherique — CHANGES_REQUESTED** [review]
### Great start @frankopkusianwar :clap: :clap: However, some methods need further refactoring in order to get approval for this project. Check :arrow_down: 

- **[REQUIRED]** Check the code comments below, they should help you refactor your methods :gear: 

### That's it :relaxed: When finished, kindly resubmit another code review request :pray: 

Happy coding :v: 
**_Ahmed Mahfoudh_**

File: enumerable_custom.rb [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    return check_all_args(arg) unless arg.nil?
+
+    condition = true
+    if block_given?
+      my_each { |item| condition = false if yield(item) == false }
+    else
+      my_each { |item| condition = false if item.nil? || item == false }
+    end
+    condition
+  end
```
Comment (shshamim63) [inline enumerable_custom.rb]: `my_all?` identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2)) 
- when a class is passed as an argument returns true if all of the collection is a member of such class
```ruby
[1, 2, 3].my_all?(Integer) #should return true
[1, 2, nil].my_all?(Integer) #should return false
```
- when a Regex is passed as an argument returns true if all of the collection matches the Regex
```ruby
%w[dog door rod blade].my_all?(/d/)   #should return true.
%w[dog door rod blade].my_all?(/t/)   #should return false.
```

File: enumerable_custom.rb [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    number
+  end
+
+  def my_map(proc = nil)
+    return to_enum(:my_map) unless block_given?
+
+    map = []
+    my_each { |item| map << proc.call(item) }
+    map
+  end
```
Comment (shshamim63) [inline enumerable_custom.rb]: `my_map` identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- returns a new array with the results of running block once for every element in enum
```ruby
block = proc { |num| num < (1 + 10) / 2 }
[1, 2 ,3].my_map(&block)   #should return [1, 2, 3]
[6, 7 ,5].my_map { |num| num > 5 }   #should also return [6, 7]
```

File: enumerable_custom.rb [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+      my_each { |item| total = yield(total, item) }
+      total
+    elsif args[0].is_a?(Symbol)
+      my_each { |item| reducer = reducer ? reducer.send(args[0], item) : item }
+      reducer
+    elsif args[0] && args[1].is_a?(Symbol)
+      my_each { |item| reducer = reducer ? reducer.send(args[1], item) : item }
+      reducer
+    end
+  end
```
Comment (shshamim63) [inline enumerable_custom.rb]: `my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- combines all elements of enum by applying a binary operation, specified by a block
```ruby
[1, 2, 3].my_inject { |memo, num| memo + num } #should return 6
[1, 2, 3].my_inject(1) { |memo, num| memo + num } #should return 7
```

**Review by shshamim63 — CHANGES_REQUESTED** [review]
## Status :1234: Required Changes :1234: 
___
Hi @frankopkusianwar :wave: ,
Awesome job. You have almost covered all the changes asked by the previous reviewer. But there are a couple of issues that are still in here. Please check the comments below to resolve them.
___
### :spades: Mandatory Changes :spades: 
___
The list of methods is failing edge cases. Please fix them by following the instruction and don't forget to test them with the given samples.
- my_all?
- my_map
- my_inject
___
##### Please fix the issues mentioned above and send another review request.
___
## :herb: Happy Coding :herb: 
### Shakhawat Hossain

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Great Job :+1: 
You have made most of the suggested changes but we still need to fix one failing edge case.
- [x] my_inject combines all elements of enum by applying a binary operation, specified by a block
```ruby
array = Array.new(100){rand(0...9)}
operation = proc { |sum, n| sum + n } 
array.my_inject(&operation) == array.inject(&operation) # should return true
```

### Martin Afani

**Comment by frankopkusianwar** [comment]
> ## Great Job 
> You have made most of the suggested changes but we still need to fix one failing edge case.
> 
> * [x]  my_inject combines all elements of enum by applying a binary operation, specified by a block
> 
> ```ruby
> array = Array.new(100){rand(0...9)}
> operation = proc { |sum, n| sum + n } 
> array.my_inject(&operation) == array.inject(&operation) # should return true
> ```
> 
> ### Martin Afani

The test case for the age case above is passing when true is returned as shown in the screenshot below

test_case

**Review by whiz25 — APPROVED** [review]
## Approved :clap: 
Merge and proceed to the next project. Best of luck.

### Martin Afani
